controllers/database_secret_handler.go

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright 2021 kloeckner.i GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	kciv1alpha1 "github.com/kloeckner-i/db-operator/api/v1alpha1"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	PgPasswordKey    = "POSTGRES_PASSWORD"
+	MysqlPasswordKey = "PASSWORD"
+	PgEngine         = "postgres"
+	MysqlEngine      = "mysql"
+)
+
+/* ------ Secret Event Handler ------ */
+type secretEventHandler struct {
+	client.Client
+}
+
+func (e *secretEventHandler) Update(evt event.UpdateEvent, q workqueue.RateLimitingInterface) {
+
+	logrus.Info("Secret Update Event processing started")
+
+	switch v := evt.ObjectNew.(type) {
+
+	default:
+		logrus.Error("Secret Update Event error! Unknown object: type=", v.GetObjectKind(), ", name=", evt.ObjectNew.GetNamespace(), "/", evt.ObjectNew.GetName())
+		return
+
+	case *corev1.Secret:
+		secretNew := evt.ObjectNew.(*corev1.Secret)
+		// find database object
+		databases, dbNames, _ := getDatabasesBySecret(e.Client, types.NamespacedName{
+			Name:      evt.ObjectNew.GetName(),
+			Namespace: evt.ObjectNew.GetNamespace(),
+		})
+
+		numDatabases := len(databases)
+		if numDatabases == 0 {
+			logrus.Error("Secret Update Event error! Could not find Database resource related to the Secret: secret=", secretNew.Namespace, "/", secretNew.Name)
+			return
+		}
+		if numDatabases > 1 {
+			// We do not allow using the same Secret resource for several Database resources!
+			logrus.Warning("Secret Update Event error! Multiple Database resources related to the same Secret: secret=", secretNew.Namespace, "/", secretNew.Name, ", dbNames=", dbNames)
+		}
+
+		database := databases[0]
```

Comment:
For the case that same secret is used for multiple databases, from here it has to be iterate process.